node/bft/src/primary.rs

```diff
@@ -141,14 +129,53 @@ impl<N: Network> Primary<N> {
             ledger,
             workers: Arc::from(vec![]),
             bft_sender: Default::default(),
-            proposed_batch: Arc::new(RwLock::new(proposed_batch)),
+            proposed_batch: Default::default(),
             latest_proposed_batch_timestamp: Default::default(),
-            signed_proposals: Arc::new(RwLock::new(signed_proposals)),
+            signed_proposals: Default::default(),
             handles: Default::default(),
-            propose_lock: Arc::new(TMutex::new(latest_certificate_round)),
+            propose_lock: Default::default(),
         })
     }
 
+    /// Load the proposal cache file and update the Primary state with the stored data.
+    async fn load_proposal_cache(&self) -> Result<()> {
+        // Fetch the signed proposals from the file system if it exists.
+        match ProposalCache::<N>::exists(self.gateway.dev()) {
+            // If the proposal cache exists, then process the proposal cache.
+            true => match ProposalCache::<N>::load(self.gateway.account().address(), self.gateway.dev()) {
+                Ok(proposal_cache) => {
+                    // Extract the proposal and signed proposals.
+                    let (latest_certificate_round, proposed_batch, signed_proposals, pending_certificates) =
+                        proposal_cache.into();
+
+                    // Write the proposed batch.
+                    *self.proposed_batch.write() = proposed_batch;
+                    // Write the signed proposals.
+                    *self.signed_proposals.write() = signed_proposals;
+                    // Writ the propose lock.
+                    *self.propose_lock.lock().await = latest_certificate_round;
+
+                    // Update the storage with the pending certificates.
+                    for certificate in pending_certificates {
```

Comment:
Can this be sped up @ljedrz ?

Reply:
I don't see anything obvious, but I'll investigate it further.